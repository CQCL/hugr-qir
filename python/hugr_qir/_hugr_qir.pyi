def cli(*args: str) -> None: ...
def opt_level_choices() -> list[str]: ...
def compile_target_choices() -> list[str]: ...
def output_format_choices() -> list[str]: ...
