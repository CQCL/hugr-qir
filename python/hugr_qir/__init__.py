def main_cli():
    from ._hugr_qir import cli
    import sys
    cli(*sys.argv)
