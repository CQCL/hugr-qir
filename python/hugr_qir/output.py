from base64 import b64encode
from enum import Enum
from pathlib import Path

from llvmlite.binding import (  # type: ignore[import-untyped]
    create_context,
    parse_assembly,
)


class OutputFormat(Enum):
    LLVMIR = "llvmir"
    BITCODE = "bitcode"
    BASE64 = "base64"


def get_write_mode(out_format: OutputFormat | None, file_path: Path | None) -> str:
    if out_format and out_format in [OutputFormat.BITCODE, OutputFormat.BASE64]:
        return "wb"
    if file_path and file_path.suffix not in [".ll", ".asm"]:
        return "wb"
    return "w"


def ir_string_to_output_format(qir_ir: str, output_format: OutputFormat) -> str | bytes:
    match output_format:
        case OutputFormat.LLVMIR:
            return qir_ir
        case OutputFormat.BITCODE:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            return module.as_bitcode()
        case OutputFormat.BASE64:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            qir_bitcode = module.as_bitcode()
            return b64encode(qir_bitcode).decode("utf-8")
    errmsg = "Unrecognized output format"
    raise ValueError(errmsg)
