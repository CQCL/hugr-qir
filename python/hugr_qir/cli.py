"""Cli for hugr-qir."""

import tempfile
from importlib.metadata import version
from pathlib import Path

import click
from quantinuum_qircheck import qircheck

from hugr_qir._hugr_qir import cli


@click.command(name="hugr-qir")
@click.argument("hugr_file", type=click.Path(exists=True, path_type=Path))
@click.option(
    "--validate-qir/--no-validate-qir",
    "validate",
    default=True,
    help="Whether to validate output",
)
@click.option(
    "-o",
    "--output",
    "outfile",
    type=click.File("w"),
    default="-",
    help="Name of output file (optional)",
)
@click.version_option(version=version("hugr_qir"))
def hugr_qir(validate: bool, hugr_file: Path, outfile: click.File) -> None:
    """This is the cli for converting hugr to qir."""

    options = ["-q"]
    with tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_file:
        tmp_options = [*options, "-o", temp_file.name]
        cli(str(hugr_file), *tmp_options)
        with Path.open(Path(temp_file.name)) as output:
            qir = output.read()
    if validate:
        qircheck(qir)

    if hasattr(outfile, "write"):
        outfile.write(qir)
    else:
        # this should never happen but the check for write is
        # needed to appease mypy (click doesn't handle typing well)
        err_msg = f"Could not write to object {outfile}"
        raise AttributeError(err_msg)


if __name__ == "__main__":
    hugr_qir()
