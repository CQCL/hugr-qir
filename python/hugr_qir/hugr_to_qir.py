import tempfile
from pathlib import Path

from .cli import hugr_qir_impl


def hugr_to_qir(hugr: bytes, validate_qir: bool = True) -> str:
    """A function for converting hugr to qir."""
    with (
        tempfile.NamedTemporaryFile(delete=True, suffix=".hugr") as temp_infile,
        tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_outfile,
    ):
        with Path.open(Path(temp_infile.name), "wb") as cli_input:
            cli_input.write(hugr)
        with Path.open(Path(temp_outfile.name), "w") as cli_output:
            hugr_qir_impl(validate_qir, Path(cli_input.name), cli_output)
        with Path.open(Path(temp_outfile.name), "r") as cli_output:
            return cli_output.read()
