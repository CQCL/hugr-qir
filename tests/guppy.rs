use std::{env, path::Path, process::{self, Command, Stdio}};

use hugr::{package::Package, std_extensions::STD_REG, Hugr};
use hugr_llvm::inkwell;
use hugr_qir::CompileArgs;
use insta::assert_snapshot;
use itertools::Itertools as _;
use rstest::rstest;

fn capture_guppy(path: impl AsRef<Path>) -> (Hugr, String) {
    let process::Output {
        status,
        stdout,
        stderr,
    } = Command::new(".devenv/state/venv/bin/python")
        .arg(path.as_ref())
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();
    eprintln!("{}", std::str::from_utf8(&stdout).unwrap());
    eprintln!("{}", std::str::from_utf8(&stderr).unwrap());
    assert!(status.success());
    let hugr = Package::from_json_reader(stdout.as_slice(), &STD_REG)
        .unwrap()
        .modules
        .into_iter()
        .exactly_one()
        .unwrap();
    let stderr = String::from_utf8(stderr).unwrap();
    (hugr, stderr)
}

fn compile(hugr: &mut Hugr) -> String {
    let args = {
        let mut a = CompileArgs::default();
        a.validate = true;
        a
    };
    let context = inkwell::context::Context::create();
    let module = args.compile(hugr, &context).unwrap();
    module.to_string()
}

// This generates a test case for each file matching 'guppy_examples/**/*.py'
//
// For now we snapshot test the stderr output of executing the python file,
// and the LLVM IR output by hugr-qir.
//
// If we get problems with new test cases not being run in dev environments we might consider
// adding a `build.rs` as in:
// <https://docs.rs/rstest/latest/rstest/attr.rstest.html#files-path-as-input-arguments>
#[rstest]
fn guppy_examples(
    #[base_dir = "guppy_examples"]
    #[files("**/*.py")]
    file: impl AsRef<Path>,
) {
    let file2 = file.as_ref();
    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file2.file_stem().unwrap().to_str().unwrap());
    settings.bind(|| {
        let (mut hugr, stderr) = capture_guppy(file);
        assert_snapshot!("llvmir", compile(&mut hugr));
        assert_snapshot!("stderr", stderr);
    });
}
