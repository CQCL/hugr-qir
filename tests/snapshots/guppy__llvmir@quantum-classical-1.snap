---
source: tests/guppy.rs
expression: compile(&mut hugr)
---
; ModuleID = 'hugr-qir'
source_filename = "hugr-qir"

%QUBIT = type opaque
%RESULT = type opaque

@0 = private unnamed_addr constant [37 x i8] c"Expected variant 1 but got variant 0\00", align 1
@prelude.panic_template = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@1 = private unnamed_addr constant [37 x i8] c"Expected variant 1 but got variant 0\00", align 1
@prelude.panic_template.1 = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@2 = private unnamed_addr constant [2 x i8] c"0\00", align 1

define void @__hugr__.main.1() {
alloca_block:
  %"20_0" = alloca {}, align 8
  %"35_0" = alloca { i1, %QUBIT* }, align 8
  %"24_0" = alloca { i1, %QUBIT* }, align 8
  %"59_0" = alloca double, align 8
  %"57_0" = alloca double, align 8
  %"55_0" = alloca double, align 8
  %"51_0" = alloca double, align 8
  %"49_0" = alloca double, align 8
  %"47_0" = alloca double, align 8
  %"36_0" = alloca %QUBIT*, align 8
  %"0" = alloca %QUBIT*, align 8
  %"41_0" = alloca { i32, i8* }, align 8
  %"42_0" = alloca %QUBIT*, align 8
  %"05" = alloca %QUBIT*, align 8
  %"44_0" = alloca %QUBIT*, align 8
  %"60_0" = alloca %QUBIT*, align 8
  %"61_0" = alloca %QUBIT*, align 8
  %"64_0" = alloca i1, align 1
  %"25_0" = alloca %QUBIT*, align 8
  %"015" = alloca %QUBIT*, align 8
  %"30_0" = alloca { i32, i8* }, align 8
  %"31_0" = alloca %QUBIT*, align 8
  %"019" = alloca %QUBIT*, align 8
  %"33_0" = alloca %QUBIT*, align 8
  %"52_0" = alloca %QUBIT*, align 8
  %"53_0" = alloca %QUBIT*, align 8
  %"62_0" = alloca i1, align 1
  %"19_0" = alloca {}, align 8
  %"14_0" = alloca {}, align 8
  %"12_0" = alloca {}, align 8
  %"65_0" = alloca i1, align 1
  %"63_0" = alloca i1, align 1
  %"17_0" = alloca i1, align 1
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  br label %0

0:                                                ; preds = %entry_block
  %1 = call %QUBIT* @__quantum__rt__qubit_allocate()
  %2 = insertvalue { i1, %QUBIT* } { i1 true, %QUBIT* poison }, %QUBIT* %1, 1
  store { i1, %QUBIT* } %2, { i1, %QUBIT* }* %"35_0", align 8
  %3 = call %QUBIT* @__quantum__rt__qubit_allocate()
  %4 = insertvalue { i1, %QUBIT* } { i1 true, %QUBIT* poison }, %QUBIT* %3, 1
  store { i1, %QUBIT* } %4, { i1, %QUBIT* }* %"24_0", align 8
  store double 0xBFF921FB54442D18, double* %"59_0", align 8
  store double 0x3FF921FB54442D18, double* %"57_0", align 8
  store double 0x400921FB54442D18, double* %"55_0", align 8
  store double 0xBFF921FB54442D18, double* %"51_0", align 8
  store double 0x3FF921FB54442D18, double* %"49_0", align 8
  store double 0x400921FB54442D18, double* %"47_0", align 8
  %"35_01" = load { i1, %QUBIT* }, { i1, %QUBIT* }* %"35_0", align 8
  %5 = extractvalue { i1, %QUBIT* } %"35_01", 0
  switch i1 %5, label %6 [
    i1 true, label %7
  ]

6:                                                ; preds = %0
  br label %cond_36_case_0

7:                                                ; preds = %0
  %8 = extractvalue { i1, %QUBIT* } %"35_01", 1
  store %QUBIT* %8, %QUBIT** %"05", align 8
  br label %cond_36_case_1

9:                                                ; preds = %30
  ret void

cond_36_case_0:                                   ; preds = %6
  store { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @0, i32 0, i32 0) }, { i32, i8* }* %"41_0", align 8
  %"41_03" = load { i32, i8* }, { i32, i8* }* %"41_0", align 8
  %10 = extractvalue { i32, i8* } %"41_03", 0
  %11 = extractvalue { i32, i8* } %"41_03", 1
  %12 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template, i32 0, i32 0), i32 %10, i8* %11)
  call void @abort()
  store %QUBIT* null, %QUBIT** %"42_0", align 8
  %"42_04" = load %QUBIT*, %QUBIT** %"42_0", align 8
  store %QUBIT* %"42_04", %QUBIT** %"0", align 8
  br label %cond_exit_36

cond_36_case_1:                                   ; preds = %7
  %"06" = load %QUBIT*, %QUBIT** %"05", align 8
  store %QUBIT* %"06", %QUBIT** %"44_0", align 8
  %"44_07" = load %QUBIT*, %QUBIT** %"44_0", align 8
  store %QUBIT* %"44_07", %QUBIT** %"0", align 8
  br label %cond_exit_36

cond_exit_36:                                     ; preds = %cond_36_case_1, %cond_36_case_0
  %"02" = load %QUBIT*, %QUBIT** %"0", align 8
  store %QUBIT* %"02", %QUBIT** %"36_0", align 8
  %"36_08" = load %QUBIT*, %QUBIT** %"36_0", align 8
  %"57_09" = load double, double* %"57_0", align 8
  %"59_010" = load double, double* %"59_0", align 8
  call void @__quantum__qis__phasedx__body(double %"57_09", double %"59_010", %QUBIT* %"36_08")
  store %QUBIT* %"36_08", %QUBIT** %"60_0", align 8
  %"60_011" = load %QUBIT*, %QUBIT** %"60_0", align 8
  %"55_012" = load double, double* %"55_0", align 8
  call void @__quantum__qis__rz__body(double %"55_012", %QUBIT* %"60_011")
  store %QUBIT* %"60_011", %QUBIT** %"61_0", align 8
  %"61_013" = load %QUBIT*, %QUBIT** %"61_0", align 8
  %13 = call %RESULT* @__quantum__qis__m__body(%QUBIT* %"61_013")
  %14 = call i1 @__quantum__qis__read_result__body(%RESULT* %13)
  %15 = select i1 %14, i1 true, i1 false
  store i1 %15, i1* %"64_0", align 1
  %"24_014" = load { i1, %QUBIT* }, { i1, %QUBIT* }* %"24_0", align 8
  %16 = extractvalue { i1, %QUBIT* } %"24_014", 0
  switch i1 %16, label %17 [
    i1 true, label %18
  ]

17:                                               ; preds = %cond_exit_36
  br label %cond_25_case_0

18:                                               ; preds = %cond_exit_36
  %19 = extractvalue { i1, %QUBIT* } %"24_014", 1
  store %QUBIT* %19, %QUBIT** %"019", align 8
  br label %cond_25_case_1

cond_25_case_0:                                   ; preds = %17
  store { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @1, i32 0, i32 0) }, { i32, i8* }* %"30_0", align 8
  %"30_017" = load { i32, i8* }, { i32, i8* }* %"30_0", align 8
  %20 = extractvalue { i32, i8* } %"30_017", 0
  %21 = extractvalue { i32, i8* } %"30_017", 1
  %22 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template.1, i32 0, i32 0), i32 %20, i8* %21)
  call void @abort()
  store %QUBIT* null, %QUBIT** %"31_0", align 8
  %"31_018" = load %QUBIT*, %QUBIT** %"31_0", align 8
  store %QUBIT* %"31_018", %QUBIT** %"015", align 8
  br label %cond_exit_25

cond_25_case_1:                                   ; preds = %18
  %"020" = load %QUBIT*, %QUBIT** %"019", align 8
  store %QUBIT* %"020", %QUBIT** %"33_0", align 8
  %"33_021" = load %QUBIT*, %QUBIT** %"33_0", align 8
  store %QUBIT* %"33_021", %QUBIT** %"015", align 8
  br label %cond_exit_25

cond_exit_25:                                     ; preds = %cond_25_case_1, %cond_25_case_0
  %"016" = load %QUBIT*, %QUBIT** %"015", align 8
  store %QUBIT* %"016", %QUBIT** %"25_0", align 8
  %"25_022" = load %QUBIT*, %QUBIT** %"25_0", align 8
  %"49_023" = load double, double* %"49_0", align 8
  %"51_024" = load double, double* %"51_0", align 8
  call void @__quantum__qis__phasedx__body(double %"49_023", double %"51_024", %QUBIT* %"25_022")
  store %QUBIT* %"25_022", %QUBIT** %"52_0", align 8
  %"52_025" = load %QUBIT*, %QUBIT** %"52_0", align 8
  %"47_026" = load double, double* %"47_0", align 8
  call void @__quantum__qis__rz__body(double %"47_026", %QUBIT* %"52_025")
  store %QUBIT* %"52_025", %QUBIT** %"53_0", align 8
  %"53_027" = load %QUBIT*, %QUBIT** %"53_0", align 8
  %23 = call %RESULT* @__quantum__qis__m__body(%QUBIT* %"53_027")
  %24 = call i1 @__quantum__qis__read_result__body(%RESULT* %23)
  %25 = select i1 %24, i1 true, i1 false
  store i1 %25, i1* %"62_0", align 1
  store {} undef, {}* %"20_0", align 1
  store {} undef, {}* %"19_0", align 1
  store {} undef, {}* %"14_0", align 1
  store {} undef, {}* %"12_0", align 1
  %"64_028" = load i1, i1* %"64_0", align 1
  %26 = select i1 %"64_028", i1 true, i1 false
  store i1 %26, i1* %"65_0", align 1
  %"62_029" = load i1, i1* %"62_0", align 1
  %27 = select i1 %"62_029", i1 true, i1 false
  store i1 %27, i1* %"63_0", align 1
  %"63_030" = load i1, i1* %"63_0", align 1
  %"65_031" = load i1, i1* %"65_0", align 1
  %28 = xor i1 %"63_030", %"65_031"
  %29 = select i1 %28, i1 true, i1 false
  store i1 %29, i1* %"17_0", align 1
  %"17_032" = load i1, i1* %"17_0", align 1
  call void @__quantum__rt__bool_record_output(i1 %"17_032", i8* getelementptr inbounds ([2 x i8], [2 x i8]* @2, i32 0, i32 0))
  %"20_033" = load {}, {}* %"20_0", align 1
  store {} %"20_033", {}* %"20_0", align 1
  %"20_034" = load {}, {}* %"20_0", align 1
  switch i1 false, label %30 [
  ]

30:                                               ; preds = %cond_exit_25
  br label %9
}

declare %QUBIT* @__quantum__rt__qubit_allocate()

declare i32 @printf(i8*, ...)

declare void @abort()

declare void @__quantum__qis__phasedx__body(double, double, %QUBIT*)

declare void @__quantum__qis__rz__body(double, %QUBIT*)

declare %RESULT* @__quantum__qis__m__body(%QUBIT*)

declare i1 @__quantum__qis__read_result__body(%RESULT*)

declare void @__quantum__rt__bool_record_output(i1, i8*)
